src/replication.c

```diff
@@ -3390,3 +3416,192 @@ void replicationStartPendingFork(void) {
+    if (server.failover_end_time <= now) {
+        if (server.masterhost) {
+            /* We attempted to failover but it did not complete in time so
+             * become a master again to undo the failover. */
+            serverLog(LL_NOTICE,
+                "FAILOVERTO %s:%d attempted, but sync was unsuccessful.",
+                server.target_replica_host, server.target_replica_port);
+            replicationUnsetMaster();
+            goto cron_complete;
+        } else if (server.force_failover) {
+            /* If timeout has expired force a failover if requested. */
+            replicationSetMaster(server.target_replica_host,
+                server.target_replica_port);
+            serverLog(LL_NOTICE,
+                "FAILOVERTO %s:%d expired, but forcing failover.",
+                server.target_replica_host, server.target_replica_port);
+            /* Extend failover time to allow replication to complete, then reset
+             * the force flag so progress will be checked during the extended
+             * duration. */
+            server.failover_end_time += FAILOVERTO_TIMEOUT;
+            server.force_failover = 0;
+            return;
+        } else {
+            /* Failover did not attempt or succeed. */
+            serverLog(LL_NOTICE,"FAILOVERTO %s:%d was unsuccessful.",
+                server.target_replica_host, server.target_replica_port);
+            goto cron_complete;
+        }
+    }
+
+    /* If offsets are equal and we havn't started failover yet, do so */
+    if (server.target_replica_host) {
+        client *replica = findReplica(server.target_replica_host,
+            server.target_replica_port);
+        if (replica && (replica->repl_ack_off == server.master_repl_offset)) {
+            /* Setup replication and extend failover time to allow replication
+             * time to complete. */
+            replicationSetMaster(server.target_replica_host,
+                server.target_replica_port);
+            server.failover_end_time = now + FAILOVERTO_TIMEOUT;
+        }
+    } else {
+        listIter li;
+        listNode *ln;
+
+        listRewind(server.slaves,&li);
+        /* Find any replica that has matched our repl_offset */
+        while((ln = listNext(&li))) {
+            client *replica = ln->value;
+            if (replica &&
+                    (replica->repl_ack_off == server.master_repl_offset)) {
+                char ip[NET_IP_STR_LEN], *replicaip = replica->slave_ip;
+
+                if (replicaip[0] == '\0') {
+                    if (connPeerToString(replica->conn,ip,sizeof(ip),NULL) == -1)
+                        continue;
+                    replicaip = ip;
+                }
+
+                /* Setup replication and extend failover time to allow
+                 * replication time to complete. */
+                replicationSetMaster(replicaip, replica->slave_listening_port);
+                server.failover_end_time = now + FAILOVERTO_TIMEOUT;
```

Comment:
A nitpick, I think a clearer implementation would be to identify the replica in the conditional block, and consolidate the replicationSetMastercode + comment after.

I think we also need to break here, the current code will setMaster if multiple replicas are caught up.

Reply:
added a break, but the consolidation was a little tricky so i did not do that